### PR TITLE
Bandit

### PR DESCRIPTION
**Pull request by oliver-sanders**
~built on #4343~

Add the Bandit static security analyser to GH actions.

### CODE REVIEW

File: .bandit [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# NOTE: At present Bandit does *not* automatically source the .bandit
+# configuration file, invoke like so:
+#   $ bandit -r --ini .bandit
+# For development it may be convenient to use flake8-bandit.
```
Comment (oliver-sanders) [inline .bandit:4]: :point_up:

File: .bandit [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+#       * The purpose of cylc is to run user defined code therefore the use
+#         of subprocesses is unavoidable.
+# B607: Starting a process with a partial executable path
+#       * Ignored as Cylc needs to be able to call out to executables which
+#         may have to be installed into environments so we cannot specify
+#         absolute paths. In some cases this may be required for portability.
+#       * Users can modify their $PATH to get Cylc to call malicious scripts,
+#         however, the purpose of Cylc Flow is to run user-defined code making
+#         this a moot point, note all code is run as the user. Cylc Flow does
+#         *not* provide multi-user functionality.
```
Comment (oliver-sanders) [inline .bandit:20]: @hjoliver could you review these lines please.

**Comment by datamel** [comment]
Maybe it would be worth adding ZAP (https://www.zaproxy.org/) to the UI for dynamic OWASP analysis? From the looks of things its a little bit of an effort to get setup but I could check it out next training day if it would be helpful.

**Comment by kinow** [comment]
> Maybe it would be worth adding ZAP (https://www.zaproxy.org/) to the UI for dynamic OWASP analysis? From the looks of things its a little bit of an effort to get setup but I could check it out next training day if it would be helpful.

I used it with Jenkins, we had it enabled for a Java project at NIWA for a while. Works well, but requires devs to create the tests and update when necessary, but it was ~4 years ago so things might be simpler (there might even exist a GH actions out there?). I'm keen to use it if others want it too.

**Comment by oliver-sanders** [comment]
Yeah, create a cylc-ui/uis issue for it.

It might be possible to [automate ZAP tests with Cypress](https://github.com/jverhoelen/security-tests-with-owasp-zap-cypress) which would make the process less manual (although it looks very fiddly at the moment).

**Comment by datamel** [comment]
> Yeah, create a cylc-ui/uis issue for it.
> 
> It might be possible to [automate ZAP tests with Cypress](https://github.com/jverhoelen/security-tests-with-owasp-zap-cypress) which would make the process less manual (although it looks very fiddly at the moment).

Will do, I know the implementation of it is not necessarily a quick job but possibly worth the time investment. I'll raise an issue.

File: .bandit (hunk 2 above)
Comment (hjoliver) [reply .bandit:20]: Looks great!  (Sorry I missed this ping 6 days ago).

**Comment by hjoliver** [comment]
Seems to be some genuine functional test failures.

**Comment by oliver-sanders** [comment]
Found a rogue quote in an SQL statement, fixed and rebased, :crossed_fingers:.

**Comment by oliver-sanders** [comment]
Only running bandit on Python 3.7 because of https://github.com/PyCQA/bandit/issues/658

**Review by datamel — APPROVED** [review]
Nice to have. Thanks @oliver-sanders.
I have read the code and checked out the branch, bandit is running smoothly locally and remotely.

**Review by MetRonnie — APPROVED** [review]

**Pull request merged**